File: linked_lists/reverse_linked_list/c++/pair_sum/pair_sum.cpp
// pair_sum.cpp
#include "pair_sum.hpp"
#include "node.hpp"

int pairSum(const ListNode::Ptr &head)
{
    // Two nodes only
    if (head->next->next->next == nullptr)
    {
        return head->next->data + head->next->next->data;
    }

    // Find the middle of the linked list
    ListNode::Ptr fastMiddle = head;
    ListNode::Ptr slowMiddle = head;
    int length = 0;
    while (fastMiddle != nullptr && fastMiddle->next != nullptr)
    {
        fastMiddle = fastMiddle->next->next;
        slowMiddle = slowMiddle->next;
        length += 1;
    }
    length *= 2;

    // Reverse the second half of the list
    ListNode::Ptr prev = nullptr;
    ListNode::Ptr lastNodeFirstHalf = slowMiddle;
    ListNode::Ptr lastNodeSecondHalf = fastMiddle;
    ListNode::Ptr current = slowMiddle->next;
    while (current != nullptr)
    {
        // Keep a refernce to the next node
        ListNode::Ptr nextNode = current->next; // Reverse the current node to point to the previous node
        current->next = prev;
        // The current node should now become the previous of the next current node
        prev = current;
        // Move the current pointer to the next node
        current = nextNode;
    }
    // To connect the first half with the reversed second half
    lastNodeFirstHalf->next = lastNodeSecondHalf;

    // Find maximum twin sum
    ListNode::Ptr fastSum = head;
    ListNode::Ptr slowSum = head;
    for (int i = 0; i < length / 2; i++)
    {
        fastSum = fastSum->next;
    }
    int maxSum = 0;
    while (fastSum != nullptr && fastSum->next != nullptr)
    {
        maxSum = std::max(fastSum->next->data + slowSum->next->data, maxSum);
        fastSum = fastSum->next;
        slowSum = slowSum->next;
    }
    return maxSum;
}

File: linked_lists/reverse_linked_list/python3/pair_sum.py
from node import ListNode
from singly import SinglyLinkedList


def pair_sum(head: ListNode) -> int:
    """
    Given the `head` of a linked list with even length, return the maximum twin sum of the linked list.

    Parameters
    ----------
    head : ListNode
        The sentinel head of the linked list

    Returns
    -------
    int
        The maximum twin sum
    """
    # Two node case
    if head.next.next.next is None:
        return head.next.data + head.next.next.data

    # Find middle: head -> a -> b -> [c] -> d -> e -> f
    fast = head
    slow = head
    length = 0
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
        length += 1
    # Get the full length
    length *= 2

    # Reverse the second half of the list to head -> a -> b -> c -> f -> e -> d
    prev = None
    last_node_second_half = fast
    last_node_first_half = slow
    current = slow.next
    while current:
        # Keep a reference to the next node
        next_node = current.next
        # Reverse the direction
        current.next = prev
        # The current node should now be considered the previous node
        prev = current
        # Keep the loop going
        current = next_node
    # To connect the first and reversed second half
    last_node_first_half.next = last_node_second_half

    # All pairs are (length / 2) steps apart: [a + f], [b + e], [c + d]
    fast = head
    slow = head
    # Move fast (length / 2) steps head
    for _ in range(length // 2):
        fast = fast.next
    max_sum = 0
    while fast and fast.next:
        # Since slow and fast are initialized to sentinal head, add their next nodes
        max_sum = max(slow.next.data + fast.next.data, max_sum)
        fast = fast.next
        slow = slow.next

    return max_sum


def main() -> int:
    data_arrays = [[5, 4, 2, 1], [4, 2, 2, 3], [1, 2], [7, 12, 27, 17, 77, 27]]
    for data_array in data_arrays:
        nodes = (ListNode(data=data) for data in data_array)
        sll = SinglyLinkedList()
        for node in nodes:
            sll.add_to_end(node)

        print("Before reversing")
        sll.display()
        max_sum = pair_sum(sll.head)
        print("After reversing")
        sll.display()
        print(f"The maximum twin sum is {max_sum}")
        print("\n")

    return 0


if __name__ == "__main__":
    main()
